spark/src/test/scala/org/apache/spark/sql/delta/CloneTableScalaSuite.scala

```diff
@@ -38,12 +45,113 @@ class CloneTableScalaSuite extends CloneTableSuiteBase
+      table.cloneAtTimestamp(timestampAsOf.get,
+        target, isShallow, replace = isReplace, tableProperties)
     } else {
-      table.clone(target, isReplace, tableProperties)
+      table.clone(target, isShallow, replace = isReplace, tableProperties)
     }
   }
   // scalastyle:on argcount
+
+  testAllClones("cloneAtVersion API") { (source, target, isShallow) =>
+    spark.range(5).write.format("delta").save(source)
+    spark.range(5).write.format("delta").mode("append").save(source)
+    spark.range(5).write.format("delta").mode("append").save(source)
+
+    val sourceTbl = io.delta.tables.DeltaTable.forPath(source)
+    assert(spark.read.format("delta").load(source).count() === 15)
+
+    runAndValidateClone(source, target, isShallow, sourceVersion = Some(0)) {
+      () => {
+        sourceTbl.cloneAtVersion(0, target, isShallow)
+      }
+    }
+  }
+
+  test("deep clone not supported yet") {
+    withSourceTargetDir { (source, clone) =>
+      assert(intercept[DeltaIllegalArgumentException] {
+        val df1 = Seq(1, 2, 3, 4, 5).toDF("id").withColumn("part", 'id % 2)
+        val df2 = Seq(8, 9, 10).toDF("id").withColumn("part", 'id % 2)
+        df1.write.format("delta").partitionBy("part").mode("append").save(source)
+        df2.write.format("delta").mode("append").save(source)
+
+        runAndValidateClone(source, clone, isShallow = false)()
+      }.getErrorClass === "DELTA_UNSUPPORTED_DEEP_CLONE")
+    }
+  }
+
+  testAllClones("clone API") { (source, target, isShallow) =>
+    spark.range(5).write.format("delta").save(source)
+    spark.range(5).write.format("delta").mode("append").save(source)
+    spark.range(5).write.format("delta").mode("append").save(source)
+
+    val sourceTbl = io.delta.tables.DeltaTable.forPath(source)
+    assert(spark.read.format("delta").load(source).count() === 15)
+
+    runAndValidateClone(source, target, isShallow) {
+      () => {
+        sourceTbl.clone(target, isShallow)
+      }
+    }
+  }
+
+  testAllClones("cloneAtTimestamp API") { (source, target, isShallow) =>
+    spark.range(5).write.format("delta").save(source)
+    spark.range(5).write.format("delta").mode("append").save(source)
+    spark.range(5).write.format("delta").mode("append").save(source)
+
+    val sourceTbl = io.delta.tables.DeltaTable.forPath(source)
+    assert(spark.read.format("delta").load(source).count() === 15)
+
+    val desiredTime = "1996-01-12"
+
+    val format = new java.text.SimpleDateFormat("yyyy-MM-dd")
+    val time = format.parse(desiredTime).getTime
+
+    val path = new Path(source + "/_delta_log/00000000000000000000.json")
+    // scalastyle:off deltahadoopconfiguration
+    val fs = path.getFileSystem(spark.sessionState.newHadoopConf())
+    // scalastyle:on deltahadoopconfiguration
+    fs.setTimes(path, time, 0)
+    if (coordinatedCommitsEnabledInTests) {
+      InCommitTimestampTestUtils.overwriteICTInDeltaFile(
```

Comment:
You can provide a clock to the Delta log I believe and that may allow achieving the same effect.
Anyway, if this method works then no need to look further